groups/views.py

```diff
@@ -34,6 +41,44 @@ def dispatch(self, request, *args, **kwargs):
         return super(GroupDetail, self).dispatch(request, *args, **kwargs)
 
 
+class SubscribeBase(SingleObjectMixin, FormView):
+    form_class = forms.SubscribeForm
+
+    def dispatch(self, request, *args, **kwargs):
+        self.object = self.get_object()
+        return super(SubscribeBase, self).dispatch(request, *args, **kwargs)
+
+    def get_form_kwargs(self):
+        """Pass the user to the form to check subscription state."""
+        kwargs = super(SubscribeBase, self).get_form_kwargs()
+        kwargs.update({
+            'user': self.request.user,
+            'instance': self.object,
+            'url_name': self.subscribe_url_name,
+        })
+        return kwargs
+
+    def form_valid(self, form):
+        """Subscribe or unsubscribe the request user."""
+        user = self.request.user
+
+        if form.initial['subscribe']:
```

Comment:
This needs to use `cleaned_data`, not `initial`, as otherwise this is not thread-safe.